### PR TITLE
BFS

### PR DESCRIPTION
**Pull request by tanmesh**

### CODE REVIEW

File: G-bfs.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// 4 6
+// 0 1
+// 0 2
+// 1 2
+// 2 0
+// 2 3 
+// 3 3
+// 2
+
+void BFS(int i, vector<bool>& visited, vector<int>& parent,  vector<int> graph[]) {
```
Comment (nilmish) [inline G-bfs.cpp]: parent and visited could be taken as a local variable here. No need to pass it from main. Only you need to pass graph and the starting variable. Normally people name starting variable as `source`

File: G-bfs.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	queue<int> q;
+	visited[i] = true;
+	q.push(i);
+
+	while(!q.empty()) {
+		i = q.front();
+		cout << i << " ";
+		q.pop();
+		for(auto it = graph[i].begin(); it!=graph[i].end(); ++it) {
```
Comment (nilmish) [inline G-bfs.cpp]: Nice to use `auto`

File: G-dfs.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+bool dfs(int u, vector<int>& visited, vector<int>& parent, vector<int> graph[]) {
+	visited[u]=1;
+	cout << u << " ";
+	for(int i=0; i<graph[u].size(); ++i) {
+		int v = graph[u][i];
+		parent[i] = u;
+		if(!visited[v] && dfs(v, visited, parent, graph)){
+			return true;
```
Comment (nilmish) [inline G-dfs.cpp]: What is true and false here for?

File: G-no_of_connected_components.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	cout << u << " ";
+	for(int i=0; i<graph[u].size(); ++i) {
+		int v = graph[u][i];
+		if(!visited[v]) {
+			dfs(v, visited, graph);
+		}
+	}
+}
+
+void connectedComponent(int v, vector<int> graph[]) {
```
Comment (nilmish) [inline G-no_of_connected_components.cpp]: Shouldn't this function return the count of `connectedComponent `?

File: G-no_of_connected_components.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	vector<bool> visited(v, false);
+	for(int i=0; i<v; ++i) {
+		if(!visited[i]){
+			dfs(i, visited, graph);
```
Comment (nilmish) [inline G-no_of_connected_components.cpp]: You can maintain a counter here. Every time you find a new un-visited node, increment the counter.

File: binaryHeaps.cpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
+#include <math.h>
+#include <iostream>
+using namespace std;
+
+void heapify(int* ptr, int i, int n) {
```
Comment (nilmish) [inline binaryHeaps.cpp]: Check my code I have emailed. This code can be made smaller

File: G-no_of_connected_components.cpp (hunk 4 above)
Comment (tanmesh) [reply G-no_of_connected_components.cpp]: i can do that..

File: binaryHeaps.cpp (hunk 6 above)
Comment (tanmesh) [reply binaryHeaps.cpp]: ok..

File: G-no_of_connected_components.cpp [hunk 7]
```diff
@@ -13,14 +13,25 @@ void dfs(int u, vector<bool>& visited, vector<int> graph[]) {
 	}
 }
 
-void connectedComponent(int v, vector<int> graph[]) {
+int connectedComponent(int v, vector<int> graph[]) {
 	vector<bool> visited(v, false);
+	int cnt=0, ans=0;
 	for(int i=0; i<v; ++i) {
 		if(!visited[i]){
+			++cnt;
```
Comment (nilmish) [inline G-no_of_connected_components.cpp]: What is the difference between cnt and ans?

File: G-no_of_connected_components.cpp [hunk 8]
```diff
@@ -13,14 +13,25 @@ void dfs(int u, vector<bool>& visited, vector<int> graph[]) {
...
 			dfs(i, visited, graph);
 			cout << endl;
 		}
+		else if(cnt!=0){
```
Comment (nilmish) [inline G-no_of_connected_components.cpp]: Why do you need this?

File: G-no_of_connected_components.cpp (hunk 5 above)
Comment (tanmesh) [reply G-no_of_connected_components.cpp]: done..

**Pull request merged**